### PR TITLE
Npi contact import

### PR DESCRIPTION
**Pull request by mayerm94**
Parse NPPES "endpoint" (contact) sheet for all entries with type "FHIR_URL". Add entries to npi_contact table, valid URLs are also added to fhir_endpoint table.

To test, you can run the database population script `make populatedb` and select the endpoint file included in the PR endpointmanager/cmd/nppescontactpopulator/endpoint_pfile_20050523-20200308.csv

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-164
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: endpointmanager/pkg/endpointmanager/npicontact.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	Other_Use_Description                      string
+	Content_Type                               string
+	Content_Description                        string
+	Other_Content_Description                  string
+	Location                                   *Location
+}
+
+// NPIContactStore is the interface for interacting with the storage layer that holds
+// NPIContact objects.
+type NPIContactStore interface {
```
Comment (edeyoung) [inline endpointmanager/pkg/endpointmanager/npicontact.go:30]: i think this can go away after a rebase

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+func ParseAndStoreNPIContactsFile(ctx context.Context, fname string, store *postgresql.Store) (int, error) {
+	// Provider Contact .csv downloaded from http://download.cms.gov/nppes/NPI_Files.html
+	lines, err := readContactCsv(ctx, fname)
+	if err != nil {
+		return -1, err
+	}
+	added := 0
+	// Loop through lines & turn into object
+	for i, line := range lines {
+		// break out of loop and return error if context has ended
```
Comment (edeyoung) [inline endpointmanager/pkg/nppesquerier/populatenpicontacts.go:142]: maybe add an `Info` or `Debug` printout like:

```
if added % 50 == 0 {
   log.Infof("Added %d NPI contacts\n", added)
}
```

File: endpointmanager/pkg/endpointmanager/npicontact.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	ID                                         int
+	NPI_ID                                     string
+	Endpoint_Type                              string
+	Endpoint_Type_Description                  string
+	Endpoint                                   string
+	Valid_URL                                  bool
+	Affiliation                                string
+	Endpoint_Description                       string
+	Affiliation_Legal_Business_Name            string
+	Normalized_Affiliation_Legal_Business_Name string
```
Comment (edeyoung) [inline endpointmanager/pkg/endpointmanager/npicontact.go:18]: is this used anywhere?

File: endpointmanager/pkg/endpointmanager/npicontact.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+// NPIContact represents the digitial contact information for an NPI Contact provided by the NPPES database
+type NPIContact struct {
+	ID                                         int
+	NPI_ID                                     string
+	Endpoint_Type                              string
+	Endpoint_Type_Description                  string
+	Endpoint                                   string
+	Valid_URL                                  bool
```
Comment (edeyoung) [inline endpointmanager/pkg/endpointmanager/npicontact.go:14]: remove snake case for these. should be `ValidURL`, etc.

**Comment by edeyoung** [comment]
recommend ensuring that `metadata` is not included in the url:
image

**Comment by edeyoung** [comment]
include note somewhere (perhaps even in the loader output message itself) that you expect to see errors at this time when adding urls to the url list.

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go [hunk 5]
```diff
@@ -110,8 +161,13 @@ func readContactCsv(ctx context.Context, filename string) ([][]string, error) {
 		// ok
 	}
 
+	err, newfilename := removeNestedDoubleQuotesFromCSV(filename)
```
Comment (edeyoung) [inline endpointmanager/pkg/nppesquerier/populatenpicontacts.go:164]: should this clean up/delete the created file at the end?

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go [hunk 6]
```diff
@@ -100,6 +104,53 @@ func isValidURL(url string) bool {
 	return urlmatched && !emailmatched
 }
 
+func removeNestedDoubleQuotesFromCSV(filename string) (error, string) {
```
Comment (edeyoung) [inline endpointmanager/pkg/nppesquerier/populatenpicontacts.go:107]: i see that the line you are testing in the test file originally contains nested double quotes. can you say that somewhere in the test to indicate that you are testing that this function is doing what is expected?

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go [hunk 7]
```diff
@@ -100,6 +104,53 @@ func isValidURL(url string) bool {
...
+	}
+
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+	var processedlines []string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		// TODO: This regex is only checking every other field starting with the second feld,
+		// it is capturing all of the free-text fields that include bad-data as of the March/2020 NPI Data
```
Comment (edeyoung) [inline endpointmanager/pkg/nppesquerier/populatenpicontacts.go:121]: that's convenient... but makes me wary.
it looks like this may do the reverse so you could combine both of these to cover the whole file: `(")(.*?)","`
image

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go (hunk 6 above)
Comment (mayerm94) [reply endpointmanager/pkg/nppesquerier/populatenpicontacts.go:107]: https://github.com/onc-healthit/lantern-back-end/pull/56/files#diff-082e02182700664adbe8904cd5c4256fR132 you mean like i mention here or more?

File: endpointmanager/pkg/nppesquerier/populatenpicontacts_test.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+	if data.EndpointType != "OTHERS" {
+		t.Errorf("Expected EndpointType to be %s, got %s", "OTHERS", data.EndpointType)
+	}
+	// EndpointTypeDescription
+	if data.EndpointTypeDescription != "Other URL" {
+		t.Errorf("Expected EndpointTypeDescription to be %s, got %s", "Other URL", data.EndpointTypeDescription)
+	}
+	// Endpoint
+	if data.Endpoint != "https://foobar.com/metadata/" {
+		t.Errorf("Expected Endpoint to be %s, got %s", "Salisbury", data.Endpoint)
```
Comment (bmath10) [inline endpointmanager/pkg/nppesquerier/populatenpicontacts_test.go:40]: This error message should be updated

File: endpointmanager/pkg/endpointmanager/postgresql/npicontactstore.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+	if err != nil {
+		return err
+	}
+	deleteNPIContactStatement, err = s.DB.Prepare(`
+		DELETE FROM npi_contacts
+		WHERE id=$1`)
+	if err != nil {
+		return err
+	}
+	if err != nil {
```
Comment (bmath10) [inline endpointmanager/pkg/endpointmanager/postgresql/npicontactstore.go:200]: Extra error check

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go (hunk 5 above)
Comment (mayerm94) [reply endpointmanager/pkg/nppesquerier/populatenpicontacts.go:164]: i considered that, but i had found it useful to have a copy of a "correct" CSV on hand, might also be useful in the future to aid in debugging should any problems arise. Fortunately the contact file is one of the smaller ones included in the NPPES package.

File: endpointmanager/pkg/nppesquerier/populatenpicontacts.go (hunk 7 above)
Comment (mayerm94) [reply endpointmanager/pkg/nppesquerier/populatenpicontacts.go]: I combined 2 regexes, but the regex you suggested wouldn't work because it captures the outer quotations as part of the value group.

**Comment by edeyoung** [comment]
i approve pending merge w master

**Comment by bmath10** [comment]
> i approve pending merge w master

me too!

**Comment by edeyoung** [comment]
this reveals a bug in capability querier. Added task to address this bug.
